pkg/k8s/kip.go

```diff
@@ -0,0 +1,170 @@
+			},
+			UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+				RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+					MaxUnavailable: &intstr.IntOrString{IntVal: int32(5), Type: intstr.Int},
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "kip-",
+					Namespace:    namespace,
+					Labels:       map[string]string{"daemonset": "kip"},
+				},
+				Spec: v1.PodSpec{
+					InitContainers: initContainers,
+					Containers: []v1.Container{
+						{
+							Name:  "pause",
+							Image: "k8s.gcr.io/pause:3.8",
+							Resources: v1.ResourceRequirements{
+								Requests: v1.ResourceList{
+									"cpu":               resource.MustParse(kipCPU),
+									"ephemeral-storage": resource.MustParse(kipEphemeralStorage),
+									"memory":            resource.MustParse(kipMemory),
+								},
+								Limits: v1.ResourceList{
+									"cpu":               resource.MustParse(kipCPU),
+									"ephemeral-storage": resource.MustParse(kipEphemeralStorage),
+									"memory":            resource.MustParse(kipMemory),
+								},
+							},
+						},
+					},
+					ImagePullSecrets: []v1.LocalObjectReference{
+						{
+							Name: imagePullSecret,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	dss, err := c.clientSet.AppsV1().DaemonSets(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, existing := range dss.Items {
+		if existing.Name == daemonset.Name {
+			return c.updateKIPDaemonset(ctx, daemonset)
+		}
+	}
+
+	_, err = c.clientSet.AppsV1().DaemonSets(namespace).Create(ctx, daemonset, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) createImagePullInitContainers(ctx context.Context) ([]v1.Container, error) {
+	profiles, err := c.getProfiles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	initContainers := []v1.Container{}
+	for idx, p := range profiles {
+		initContainers = append(initContainers, v1.Container{
+			Name:    fmt.Sprintf("puller-%v", idx),
+			Image:   p.KubespawnerOverride.Image,
+			Command: []string{"/bin/sh", "-c", "echo pull complete"},
+			Resources: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					"cpu":               resource.MustParse(kipCPU),
+					"ephemeral-storage": resource.MustParse(kipEphemeralStorage),
+					"memory":            resource.MustParse(kipMemory),
+				},
+				Limits: v1.ResourceList{
+					"cpu":               resource.MustParse(kipCPU),
+					"ephemeral-storage": resource.MustParse(kipEphemeralStorage),
+					"memory":            resource.MustParse(kipMemory),
+				},
+			},
+		})
+	}
+
+	return initContainers, nil
+}
+
+func (c *Client) getProfiles(ctx context.Context) ([]JupyterProfileList, error) {
+	globVals, err := c.repo.GlobalValuesGet(ctx, gensql.ChartTypeJupyterhub)
+	if err != nil {
+		return nil, err
+	}
+
+	var profiles []JupyterProfileList
+	for _, gv := range globVals {
+		if gv.Key == "singleuser.profileList" {
```

Comment:
Er det ikke bedre å hente denne verdien direkte fra databasen? I stedet for å løpe igjennom alle sammen og se om du finner den?